Sources/NIOPosix/Bootstrap.swift

```diff
@@ -1264,6 +1283,50 @@ public final class ClientBootstrap: NIOClientTCPBootstrapProtocol {
         }
     }
 
+    #if canImport(Darwin) || os(Linux)
+    public func connect(to vsockAddress: VsockAddress) throws -> EventLoopFuture<Channel> {
+        let eventLoop = group.next()
+        let channelInitializer = self.channelInitializer
+        let socket: Socket
+        let channel: SocketChannel
+        do {
+            socket = try Socket(protocolFamily: .vsock, type: .stream)
```

Comment:
Preferentially we'd have the `Channel` be responsible for doing this: that is, instead of calling `connect` here we'd have the `Channel` do it in `connect0`. This would use a user event to trigger the operation, as we do in NIOTS.

Reply:
I've pushed a commit which moves all of the logic to a user event.

There came a point where I had to bridge the ability to connect and bind to either a `SocketAddress` or a `VsockAddress`.

I ended up adding an overload of `connectSocket(to:)` in `BaseSocketChannel` and, consequently, implementations in `PipeChannel` and `DatagramChannel` which return `operationUnsupported`.

I then added a `final` function to `BaseSocketChannel` to bridge this functionality with an internal enum type.

It didn't seem as complicated with the bind side, because there connect had the non-blocking flow to handle.

Hopefully I got the right end of the stick here!